### ISSUE TITLE
vip seems to take too long time to be available in the kernel 

### ISSUE THREAD

**Issue by rdesniol** [acassen/keepalived#539]
Dear keepalived team,

On a system running ubuntu 16.04 with a specific kernel 4.4.50, i am using keepalived 1.3.2 to manage the ike daemon strongswan via the notify script.

I start keepalived via systemd. When entering in the MASTER state, the notify script launches strongswan (using also a systemctl command).

The problem is that strongswan starts the tunnel immediatly and wants to use the vip to initiate ike exchange and it fails as the vip is not yet available in the kernel. 
Here are the messages  related to this issue :

```
15:23:04 srv-test charon: 08[CFG] left nor right host is our side, assuming left=local
15:23:04 srv-test charon: 03[NET] error writing to socket: Invalid argument
```

Using IKEv2 retransmission mechanism it will retry after 4s at :15:23:**08** 
In between the vip appears in the kernel :

```
15:23:06 srv-test charon: 04[KNL] 1.1.1.3 appeared on wan
```

and the second attempt succeeds

```
15:23:08 srv-test charon: 06[IKE] retransmit 1 of request with message ID 0
15:23:08 srv-test charon: 06[NET] sending packet: from 1.1.1.3[500] to 2.1.1.3[500] (296 bytes)
```

So this behavior adds 4s in the failover process which not really expected.

Is there a way to launch the command in the notify script ONLY when the VIP is available in the kernel  ? 
Note that i do not want to rely on a sleep X inserted in notify script as it is not deterministic.

Thanks for your help;

Here are all the logs in the rigth order 

 ```
15:23:00 srv-test systemd[1]: Starting VRRP instance in network namespace vrf1 (Keepalived)...
15:23:00 srv-test Keepalived[53023]: Starting Keepalived v1.3.2 (02/06,2017)
15:23:00 srv-test Keepalived[53023]: WARNING - default user 'keepalived_script' for script execution does not exist - please create.
15:23:00 srv-test Keepalived[53023]: Opening file '/etc/keepalived/keepalived.vrf1.conf'.
15:23:00 srv-test Keepalived[53023]: Changing syslog ident to Keepalived_vrf1
15:23:00 srv-test Keepalived_vrf1[53026]: Starting VRRP child process, pid=53027
15:23:00 srv-test systemd[1]: Started VRRP instance in network namespace vrf1 (Keepalived).
15:23:00 srv-test Keepalived_vrrp_vrf1[53027]: Registering Kernel netlink reflector
15:23:00 srv-test Keepalived_vrrp_vrf1[53027]: Registering Kernel netlink command channel
15:23:00 srv-test Keepalived_vrrp_vrf1[53027]: Registering gratuitous ARP shared channel
15:23:00 srv-test Keepalived_vrrp_vrf1[53027]: Opening file '/etc/keepalived/keepalived.vrf1.conf'.
15:23:00 srv-test Keepalived_vrrp_vrf1[53027]: Unable to load ipset library - libipset.so.3: cannot open shared object file: No such file or directory
15:23:00 srv-test Keepalived_vrrp_vrf1[53027]: Using LinkWatch kernel netlink reflector...
srv-test:~$ Mar  7 15:23:00 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(wan) Entering BACKUP STATE
15:23:00 srv-test Keepalived_vrrp_vrf1[53027]: Opening script file /usr/local/bin/keepalived_notify.sh
15:23:00 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(lan) Entering BACKUP STATE
15:23:00 srv-test systemd[1]: Stopped strongSwan IPsec in network namespaces vrf1.
15:23:00 srv-test root: /usr/local/bin/keepalived_notify.sh: Run /bin/systemctl stop ipsec@vrf1
15:23:03 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(lan) Transition to MASTER STATE
15:23:03 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Group(VG_IPV4) Syncing instances to MASTER state
15:23:03 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(wan) Transition to MASTER STATE
15:23:03 srv-test Keepalived_vrrp_vrf1[53027]: Opening script file /usr/local/bin/keepalived_notify.sh
15:23:04 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(lan) Entering MASTER STATE
15:23:04 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(lan) using locally configured advertisement interval (1000 milli-sec)
15:23:04 srv-test systemd[1]: Started strongSwan IPsec in network namespaces vrf1.
15:23:04 srv-test root: /usr/local/bin/keepalived_notify.sh: Run /bin/systemctl start ipsec@vrf1
15:23:04 srv-test ip[53040]: Starting strongSwan 5.5.1 IPsec [starter]...
15:23:04 srv-test charon: 00[DMN] Starting IKE charon daemon (strongSwan 5.5.1, Linux 4.4.50-Test-9, x86_64)
15:23:04 srv-test charon: 00[CFG] loading ca certificates from '/etc/ipsec.d/cacerts'
15:23:04 srv-test charon: 00[CFG]   loaded ca certificate "C=COM, CN=Test Signing CA" from '/etc/ipsec.d/cacerts/signing-ca.crt'
15:23:04 srv-test charon: 00[CFG]   loaded ca certificate "C=COM, CN=Test Root CA" from '/etc/ipsec.d/cacerts/root-ca.crt'
15:23:04 srv-test charon: 00[CFG] loading aa certificates from '/etc/ipsec.d/aacerts'
15:23:04 srv-test charon: 00[CFG] loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
15:23:04 srv-test charon: 00[CFG] loading attribute certificates from '/etc/ipsec.d/acerts'
15:23:04 srv-test charon: 00[CFG] loading crls from '/etc/ipsec.d/crls'
15:23:04 srv-test charon: 00[CFG]   loaded crl from '/etc/ipsec.d/crls/signing-ca.crl'
15:23:04 srv-test charon: 00[CFG]   loaded crl from '/etc/ipsec.d/crls/root-ca.crl'
15:23:04 srv-test charon: 00[CFG] loading secrets from '/etc/ipsec.secrets'
15:23:04 srv-test charon: 00[CFG]   loaded ECDSA private key from '/etc/ipsec.d/private/srv.Test.eu.key'
15:23:04 srv-test charon: 00[LIB] loaded plugins: charon aesni aes rc2 sha2 shsrv md5 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp agent xcbc hmac gcm attr kernel-netlink resolve socket-default connmark stroke vici updown
15:23:04 srv-test charon: 00[LIB] dropped capabilities, running as uid 0, gid 0
15:23:04 srv-test charon: 00[JOB] spawning 16 worker threads
15:23:04 srv-test ip[53040]: charon (53056) started after 40 ms
15:23:04 srv-test charon: 08[CFG] received stroke: add connection 'conn1_1'
15:23:04 srv-test charon: 08[CFG] left nor right host is our side, assuming left=local
15:23:04 srv-test charon: 08[CFG]   loaded certificate "C=COM,  OU=Test, CN=srv.Test.eu" from 'srv.Test.eu.crt'
15:23:04 srv-test charon: 08[CFG] added configuration 'conn1_1'
15:23:04 srv-test charon: 12[CFG] received stroke: initiate 'conn1_1'
15:23:04 srv-test charon: 12[IKE] initiating IKE_SA conn1_1[1] to 20.1.1..3
15:23:04 srv-test charon: 12[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
15:23:04 srv-test charon: 12[NET] sending packet: from 10.1.1.3[500] to 20.1.1..3[500] (296 bytes)
15:23:04 srv-test charon: 03[NET] error writing to socket: Invalid argument
15:23:06 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(wan) Entering MASTER STATE
15:23:06 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(wan) using locally configured advertisement interval (1000 milli-sec)
15:23:06 srv-test charon: 04[KNL] 10.1.1.3 appeared on wan
15:23:08 srv-test charon: 06[IKE] retransmit 1 of request with message ID 0
15:23:08 srv-test charon: 06[NET] sending packet: from 10.1.1.3[500] to 20.1.1..3[500] (296 bytes)
15:23:08 srv-test charon: 12[NET] received packet: from 20.1.1..3[500] to 10.1.1.3[500] (341 bytes)
15:23:08 srv-test charon: 12[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(HASH_ALG) N(MULT_AUTH) ]
15:23:08 srv-test charon: 12[IKE] received cert request for "C=COM, CN=Test Root CA"
15:23:08 srv-test charon: 12[IKE] received cert request for "C=COM, CN=Test Signing CA"
15:23:08 srv-test charon: 12[IKE] sending cert request for "C=COM, CN=Test Signing CA"
15:23:08 srv-test charon: 12[IKE] sending cert request for "C=COM, CN=Test Root CA"
15:23:08 srv-test charon: 12[IKE] authentication of 'C=COM,  OU=Test, CN=srv.Test.eu' (myself) with ECDSA_WITH_SHA256_DER successful
15:23:08 srv-test charon: 12[IKE] sending end entity cert "C=COM,  OU=Test, CN=srv.Test.eu"
15:23:08 srv-test charon: 12[IKE] establishing CHILD_SA conn1_1
15:23:08 srv-test charon: 12[ENC] generating IKE_AUTH request 1 [ IDi CERT N(INIT_CONTACT) CERTREQ IDr AUTH SA TSi TSr N(MULT_AUTH) N(EAP_ONLY) ]
15:23:08 srv-test charon: 12[ENC] splitting IKE message with length of 1305 bytes into 2 fragments
15:23:08 srv-test charon: 12[ENC] generating IKE_AUTH request 1 [ EF(1/2) ]
15:23:08 srv-test charon: 12[ENC] generating IKE_AUTH request 1 [ EF(2/2) ]
15:23:08 srv-test charon: 12[NET] sending packet: from 10.1.1.3[500] to 20.1.1..3[500] (1252 bytes)
15:23:08 srv-test charon: 12[NET] sending packet: from 10.1.1.3[500] to 20.1.1..3[500] (118 bytes)
15:23:08 srv-test charon: 11[NET] received packet: from 20.1.1..3[500] to 10.1.1.3[500] (1134 bytes)
15:23:08 srv-test charon: 11[ENC] parsed IKE_AUTH response 1 [ IDr CERT AUTH SA TSi TSr N(AUTH_LFT) ]
15:23:08 srv-test charon: 11[IKE] received end entity cert "C=COM,  OU=Test, CN=b1.Test.eu"
15:23:08 srv-test charon: 11[CFG]   using certificate "C=COM,  OU=Test, CN=b1.Test.eu"
15:23:08 srv-test charon: 11[CFG]   using trusted intermediate ca certificate "C=COM, CN=Test Signing CA"
15:23:08 srv-test charon: 11[CFG] checking certificate status of "C=COM,  OU=Test, CN=b1.Test.eu"
15:23:08 srv-test charon: 11[CFG]   using trusted ca certificate "C=COM, CN=Test Root CA"
15:23:08 srv-test charon: 11[CFG]   reached self-signed root ca with a path length of 0
15:23:08 srv-test charon: 11[CFG]   using trusted certificate "C=COM, CN=Test Signing CA"
15:23:08 srv-test charon: 11[CFG]   crl correctly signed by "C=COM, CN=Test Signing CA"
15:23:08 srv-test charon: 11[CFG]   crl is valid: until Mar 07 17:11:03 2017
15:23:08 srv-test charon: 11[CFG]   using cached crl
15:23:08 srv-test charon: 11[CFG] certificate status is good
15:23:08 srv-test charon: 11[CFG]   using trusted ca certificate "C=COM, CN=Test Root CA"
15:23:08 srv-test charon: 11[CFG] checking certificate status of "C=COM, CN=Test Signing CA"
15:23:08 srv-test charon: 11[CFG]   using trusted certificate "C=COM, CN=Test Root CA"
15:23:08 srv-test charon: 11[CFG]   crl correctly signed by "C=COM, CN=Test Root CA"
15:23:08 srv-test charon: 11[CFG]   crl is valid: until Feb 28 17:09:24 2018
15:23:08 srv-test charon: 11[CFG]   using cached crl
15:23:08 srv-test charon: 11[CFG] certificate status is good
15:23:08 srv-test charon: 11[CFG]   reached self-signed root ca with a path length of 1
15:23:08 srv-test charon: 11[IKE] authentication of 'C=COM,  OU=Test, CN=b1.Test.eu' with ECDSA_WITH_SHA256_DER successful
15:23:08 srv-test charon: 11[IKE] IKE_SA conn1_1[1] established between 10.1.1.3[C=COM,  OU=Test, CN=srv.Test.eu]...20.1.1..3[C=COM,  OU=Test, CN=b1.Test.eu]
15:23:08 srv-test charon: 11[IKE] scheduling reauthentication in 85445s
15:23:08 srv-test charon: 11[IKE] maximum IKE_SA lifetime 85985s
15:23:08 srv-test charon: 11[IKE] CHILD_SA conn1_1{1} established with SPIs c633a0c4_i c043d4b7_o and TS 1.1.0.0/24 12.0.0.0/8 === 2.1.0.0/24 22.0.0.0/8
```

**Comment by pqarmitage** [comment]
I think the issue is around the following entries:
```
15:23:03 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(lan) Transition to MASTER STATE
15:23:03 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Group(VG_IPV4) Syncing instances to MASTER state
15:23:03 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(wan) Transition to MASTER STATE
15:23:03 srv-test Keepalived_vrrp_vrf1[53027]: Opening script file /usr/local/bin/keepalived_notify.sh
15:23:04 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(lan) Entering MASTER STATE
15:23:04 srv-test Keepalived_vrrp_vrf1[53027]: VRRP_Instance(lan) using locally configured 
advertisement interval (1000 milli-sec)
```
The problem is that there is a 1 advert interval between the `Transition to MASTER STATE` and the `Entering MASTER STATE`. The notify script is run when it logs the Transition to master state, but keepalived doesn't actually become master (i.e. adding VIPs) until the Entering master state.

I have a development branch at https://github.com/pqarmitage/keepalived/tree/fixes.quick, for which I intend to submit a pull request soon. This version has removed the 1 advert interval delay, and so the notify script should be run after the VIP has been added. It would be really helpful if you could test my development version to see if it resolves your problem.

**Comment by rdesniol** [comment]
dear pqarmitage
I confirm that i have no more the problem describeb.
I suppose that the fix found in  07a281b of dev branch "fixes.quick" will be avaiable in next 1.3.5 ?
Thanks for your help.

**Comment by pqarmitage** [comment]
@rdesniol Many thanks for testing my development branch. I don't think the fix will be added to the 1.3.x series, but in the next major release. I am hoping that won't be to far in the future.

**Comment by rdesniol** [comment]
ok thanks i close this ticket
